grudge/execution.py

```diff
@@ -150,26 +149,19 @@ def map_elementwise_max(self, op, field_expr):
     # {{{ nodal reductions
 
     def map_nodal_sum(self, op, field_expr):
-        # FIXME: Could allow array scalars
-        # FIXME: Fix CL-specific-ness
-        return sum([
-                cl.array.sum(grp_ary).get()[()]
-                for grp_ary in self.rec(field_expr)
-                ])
+        actx = self.array_context
+        return actx.np.sum(actx.np.stack([actx.np.sum(grp_ary)
+                                          for grp_ary in self.rec(field_expr)]))
```

Comment:
The `stack` here goes in the wrong direction IMO: It'll expend lots of memory bandwidth producing a temporary array that'll just get thrown away. I think the outer reductions are fine here. For `min` and `max`, you probably want two nested calls to `actx.np.{min,max}`: One two reduce the arrays, and another across arrays.

Reply:
How do you ensure the outer reduction actually knows what to do with the inner reductions (that's what I see the `actx.np.stack` doing here)? Specifically the problem across a group of arrays. AFAICT, lists and generators are something `actx.np.{sum,min,max}` do not like.

Reply:
Replied as a comment, but I don't see a problem why actx.np.stack shouldn't take a list of arrays.

Reply:
@kaushikcfd I think I misrepresented what I was trying to say. What I mean to say is this: I agree with your approach and I don't see another way to do it. The issue I'm referencing is completely removing `actx.np.stack` (which I had thought it was @inducer is suggesting, but maybe I'm wrong?). For example:
```python
actx.np.{min,max}([actx.np.{min,max}(grp_ary) for grp_ary in self.rec(field_expr)])
```
does _not _ work for the reason I described before.

Reply:
Oops thanks @thomasgibson for clarifying it out. You know what, now that you pointed it out I think one way we could do is:
```python
from functools import reduce
return reduce(lambda acc, xi: actx.np.max(acc, actx.np.max(xi)), self.rec(field_expr), -np.inf) 
```

Reply:
@thomasgibson: Is this logic being handled in your branch, or should we handle it here.

Reply:
I'll give it a try, but an initial attempt produced the following stacktrace:
```
a = -inf, b = 1.1942196769792337, out = None, queue = <pyopencl._cl.CommandQueue object at 0x7f8c7c36f7c0>

    def maximum(a, b, out=None, queue=None):
        """Return the elementwise maximum of *a* and *b*."""
    
        # silly, but functional
>       return if_positive(a.mul_add(1, b, -1, queue=queue), a, b,
                queue=queue, out=out)
E       AttributeError: 'float' object has no attribute 'mul_add'

../../pyopencl/pyopencl/array.py:2724: AttributeError
```

I'll try to see why. Basically the functions I wrote should more or less be identical to those that you've updated.

Reply:
That's pyopencl being buggy: https://github.com/inducer/pyopencl/pull/457

Reply:
FWIW, I felt this was the correct way to solve and added it here: dba33e3. Handling pyopencl comments now, to close this.

Reply:
I quick like this solution. Sounds like https://github.com/inducer/pyopencl/pull/457 needs to get merged :)

Reply:
That one's again on me.. :laughing:

Reply:
np --- It's good to know there exists a clean solution to this problem. nice job!